### ISSUE TITLE
0010518:
No ActiveSync in Windows Phone 8.1

### ISSUE THREAD

**Issue by Gloirin** [tine20/Tine-2.0-Open-Source-Groupware-and-CRM#5242]
**Reported by schoebelh on 20 Nov 2014 18:12**

**Version:** Koriander (2014.09.4)

actual i have 3 WinPhone 8.1 devices who cannot sync via activesync. This is faulty since update 2014.09.03 - before no problems. At same server are iPhone, iPad and ActiveSync devices in ActiveSync without problems!

**Additional information:** error on device:
In &#039;settings&#039; - &#039;e-mail accounts&#039; on the tine20-activesync-account: &quot;Nicht aktuell&quot; (german)

when you select the account: lot of text... Error code 85010014

accesslog:

217.84.212.xxx - hagen [20/Nov/2014:19:01:22 +0100] &quot;OPTIONS /Microsoft-Server-ActiveSync?User=hagen&amp;DeviceId=796ACD8C9822BB3C18D373282Fxxxxxx&amp;DeviceType=WP8 HTTP/1.1&quot; 200 822 &quot;-&quot; &quot;MSFT-WP/8.10.14157&quot;
217.84.212.xxx - hagen [20/Nov/2014:19:01:22 +0100] &quot;POST /Microsoft-Server-ActiveSync?jREHBBB5as2MmCK7PBjTcygvbVXsBAAAAAxxxxxx HTTP/1.1&quot; 500 309 &quot;-&quot; &quot;-&quot;
217.84.212.xxx - hagen [20/Nov/2014:19:01:23 +0100] &quot;POST /Microsoft-Server-ActiveSync?jQkHBBB5as2MmCK7PBjTcygvbVXsBAAAAAxxxxxx HTTP/1.1&quot; 500 470 &quot;-&quot; &quot;-&quot;

90.187.229.xxx - hagen [20/Nov/2014:19:01:16 +0100] &quot;POST /Microsoft-Server-ActiveSync?User=hagen&amp;DeviceId=ApplDMPLWN7Zxxxx&amp;DeviceType=iPad&amp;Cmd=Ping HTTP/1.1&quot; 200 4331 &quot;-&quot; &quot;Apple-iPad4C2/1202.410&quot;
217.84.212.xxx - hagen [20/Nov/2014:19:02:18 +0100] &quot;POST /Microsoft-Server-ActiveSync?User=hagen&amp;DeviceId=ApplDMPLWN7Zxxxx&amp;DeviceType=iPad&amp;Cmd=Sync HTTP/1.1&quot; 200 6075 &quot;-&quot; &quot;Apple-iPad4C2/1202.410&quot;
217.84.212.xxx - hagen [20/Nov/2014:19:02:18 +0100] &quot;POST /Microsoft-Server-ActiveSync?User=hagen&amp;DeviceId=ApplDMPLWN7Zxxxx&amp;DeviceType=iPad&amp;Cmd=Sync HTTP/1.1&quot; 200 11539 &quot;-&quot; &quot;Apple-iPad4C2/1202.410&quot;
...

**Comment by Gloirin** [comment]
**Comment posted by ingoratsdorf on 25 Nov 2014 10:25**

Same happens with windows 8.1 64bit with Outlook 15.
Tine20 expects POST requests in the form of

POST https://tine.spahn.me/Microsoft-Server-ActiveSync?frontend=activesync&amp;Cmd=Ping&amp;User=kekstod&amp;DeviceId=Nine7EDA6BC53456&amp;DeviceType=Android [^] HTTP/1.1

but windows 8.1 or Outlook 15 (2003) sends the following:

125.237.60.60 - xxxx [25/Nov/2014:23:08:49 +1300] &quot;POST /Microsoft-Server-ActiveSync?jRAJFBAwg8FCokZ0vYoHewQhHqg2BAAAAAALV2luZG93c01haWw= HTTP/1.1&quot; 501 566 &quot;-&quot; &quot;WindowsMail/17.5.9600.20689&quot;

note the return error code 501.

Now when we look in the tine20 log, we can see the following:

ec09a xxxx - 2014-11-25T10:08:49+00:00 CRIT (2): Syncroton_Server::_handlePost::109 command not supported:

Further looking into 

tine20/vendor/syncroton/syncroton/lib/Syncroton/Server.php, around line 109:

        $requestParameters = $this-&gt;_getRequestParameters($this-&gt;_request);
        
        if ($this-&gt;_logger instanceof Zend_Log) 
            $this-&gt;_logger-&gt;debug(__METHOD__ . &#039;::&#039; . __LINE__ . &#039; REQUEST &#039; . print_r($requestParameters, true));
        
        $className = &#039;Syncroton_Command_&#039; . $requestParameters[&#039;command&#039;];
        
        if(!class_exists($className)) {
            if ($this-&gt;_logger instanceof Zend_Log)
                $this-&gt;_logger-&gt;crit(__METHOD__ . &#039;::&#039; . __LINE__ . &quot; command not supported: &quot; . $requestParameters[&#039;command&#039;]);
            
            header(&quot;HTTP/1.1 501 not implemented&quot;);
            
            return;
        }


IN SUMMARY:
Tine20 EXPECTS a cmd=xxxxxx parameter in the POST request, but there is none, only a single string of characters, and as such, Tine20 fails with 501.

So, is that a revised ActiveSync protocol by MS?

**Comment by Gloirin** [comment]
**Comment posted by ingoratsdorf on 25 Nov 2014 10:27**

Is that somehow possibly related to #0010524 ?

**Comment by Gloirin** [comment]
**Comment posted by ingoratsdorf on 25 Nov 2014 10:58**

Okay, windows 8.1 apparently packs its requests into a base64 encoded string instead of sending it the old fashioned way:

&lt;?php

function byte_strlen($str) {
        return MBSTRING_OVERLOAD &amp; 2 ? mb_strlen($str,&#039;ascii&#039;) : strlen($str);
}

function byte_substr(&amp;$data,$offset,$len=null) {
        if ($len == null)
                return MBSTRING_OVERLOAD &amp; 2 ? mb_substr($data,$offset,byte_strlen($data),&#039;ascii&#039;) : substr($data,$offset);
        return MBSTRING_OVERLOAD &amp; 2 ? mb_substr($data,$offset,$len,&#039;ascii&#039;) : substr($data,$offset,$len);
}

function base64uri_decode($uri) {
       $uri = base64_decode($uri);
       $lenDevID = ord($uri{4});
       $lenPolKey = ord($uri{4+(1+$lenDevID)});
       $lenDevType = ord($uri{4+(1+$lenDevID)+(1+$lenPolKey)});
       $arr_ret = unpack(&quot;CProtVer/CCommand/vLocale/CDevIDLen/H&quot;.($lenDevID*2).&quot;DevID/CPolKeyLen&quot;.($lenPolKey == 4 ? &quot;/VPolKey&quot; : &quot;&quot;).&quot;/CDevTypeLen/A&quot;.($le$
       $pos = (7+$lenDevType+$lenPolKey+$lenDevID);
       $uri = byte_substr($uri,$pos);
       while (byte_strlen($uri) &gt; 0) {
           $lenToken = ord($uri{1});
           switch (ord($uri{0})) {
               case 0 : $type = &quot;AttachmentName&quot;; break;
               case 1 : $type = &quot;CollectionId&quot;; break;    //accroding to spec  20090712
               case 2 : $type = &quot;CollectionName&quot;; break;  //accroding to spec  20090712
               case 3 : $type = &quot;ItemId&quot;; break;
               case 4 : $type = &quot;LongId&quot;; break;
               case 5 : $type = &quot;ParentId&quot;; break;        //accroding to spec  20090712
               case 6 : $type = &quot;Occurrence&quot;; break;
               case 7 : $type = &quot;Options&quot;; break;
               case 8 : $type = &quot;User&quot;; break;
               default : $type = &quot;unknown&quot;.ord($uri{0}); break;
           }
           $value = unpack(&quot;CType/CLength/A&quot;.$lenToken.&quot;Value&quot;,$uri);
           $arr_ret[$type] = $value[&#039;Value&#039;];
           $pos = 2+$lenToken;
           $uri = byte_substr($uri,$pos);
       }
       return $arr_ret;
  }

$uri_decoded = base64uri_decode(&quot;jRAJFBAwg8FCokZ0vYoHewQhHqg2BAAAAAALV2luZG93c01haWw=&quot;);
print_r($uri_decoded);

results in

Array
(
    [ProtVer] =&gt; 141
    [Command] =&gt; 16
    [Locale] =&gt; 5129
    [DevIDLen] =&gt; 16
    [DevID] =&gt; 3083c142a24674bd8a077b04211ea836
    [PolKeyLen] =&gt; 4
    [PolKey] =&gt; 0
    [DevTypeLen] =&gt; 11
    [DevType] =&gt; WindowsMail
)

NOTE: I got this piece of code from egroupware and inserted my POST arguments from above.

**Comment by Gloirin** [comment]
**Comment posted by ingoratsdorf on 26 Nov 2014 10:17**

Figured it out:
The Server.php was parsing the requestURI completely wrong, see notes below.
It was expecting the URL to be
/Microsoft-Server-ActiveSync?jRQJFBAwg8FCokZ0vYoHewQhHqg2BAAAAAALV2luZG93c01haWw=
however it is delivered as
//tine20.org/Microsoft-Server-ActiveSync?frontend=activesync&amp;jRQJFBAwg8FCokZ0vYoHewQhHqg2BAAAAAALV2luZG93c01haWw=

The below fixes it, however strange URI. Rewrite rules wrong for tine20?
Should it not be:
//tine20.org/index.php?frontend=activesync&amp;jRQJFBAwg8FCokZ0vYoHewQhHqg2BAAAAAALV2luZG93c01haWw=

Although that would change nothing on the code below!



syncroton/Server.php:

    protected function _getRequestParameters(Zend_Controller_Request_Http $request)
    {
		// Request URI may have the form of
		// //tine20.org/Microsoft-Server-ActiveSync?frontend=activesync&amp;jRQJFBAwg8FCokZ0vYoHewQhHqg2BAAAAAALV2luZG93c01haWw=
		// -OR-
		// //tine20.org/Microsoft-Server-ActiveSync?frontend=activesync&amp;Cmd=Sync&amp;User=rmjones&amp;DeviceId=v140Device&amp;DeviceType=SmartPhone
		$requestParameters = substr($request-&gt;getRequestUri(), strpos($request-&gt;getRequestUri(), &#039;&amp;&#039;)+1);
		
        if (strpos($requestParameters, &#039;&amp;&#039;) === false) {
			// we have a base64 encoded request string
            $commands = array(
                0  =&gt; &#039;Sync&#039;,
                1  =&gt; &#039;SendMail&#039;,
                2  =&gt; &#039;SmartForward&#039;,
                3  =&gt; &#039;SmartReply&#039;,
                4  =&gt; &#039;GetAttachment&#039;,
                9  =&gt; &#039;FolderSync&#039;,
                10 =&gt; &#039;FolderCreate&#039;,
                11 =&gt; &#039;FolderDelete&#039;,
                12 =&gt; &#039;FolderUpdate&#039;,
                13 =&gt; &#039;MoveItems&#039;,
                14 =&gt; &#039;GetItemEstimate&#039;,
                15 =&gt; &#039;MeetingResponse&#039;,
                16 =&gt; &#039;Search&#039;,
                17 =&gt; &#039;Settings&#039;,
                18 =&gt; &#039;Ping&#039;,
                19 =&gt; &#039;ItemOperations&#039;,
                20 =&gt; &#039;Provision&#039;,
                21 =&gt; &#039;ResolveRecipients&#039;,
                22 =&gt; &#039;ValidateCert&#039;
            );
            
            $stream = fopen(&quot;php://temp&quot;, &#039;r+&#039;);
            fwrite($stream, base64_decode($requestParameters));
            rewind($stream);

**Comment by Gloirin** [comment]
**Comment posted by lkneschke on 26 Nov 2014 13:49**

Can you post the whole content from the logfile, while processing the request?

I&#039;m interessted in seeing the headers sent.

**Comment by Gloirin** [comment]
**Comment posted by schoebelh on 26 Nov 2014 16:30**

Could please exactly describe the needed log (apache, tine, debug level?)?

The changes described in ~0015048 did not solve my WP8.1 problem :-/

**Comment by Gloirin** [comment]
**Comment posted by lkneschke on 26 Nov 2014 17:47**

tine20.log is needed.

I&#039;m aware of the source of the problem, just want to be sure I fixed it correcttly. To write a test, I need the data from the log.

**Comment by Gloirin** [comment]
**Comment posted by schoebelh on 26 Nov 2014 18:19**

Log attached, hope it is the right one.

**Comment by Gloirin** [comment]
**Comment posted by lkneschke on 27 Nov 2014 06:49**

Please replace ActiveSync/Server/Http.php with attached Http.php. It should fix the problems.

**Comment by Gloirin** [comment]
**Comment posted by schoebelh on 28 Nov 2014 15:17**

Only replace Http.php or also adapt the changes from ~0015048?

**Comment by Gloirin** [comment]
**Comment posted by schoebelh on 28 Nov 2014 15:18**

When i only replace Http.php the problem with my WinPhone 8.1 devices is not solved.

**Comment by Gloirin** [comment]
**Comment posted by lkneschke on 28 Nov 2014 21:34**

There was another problem. Please also replace Tinebase/Controller.php with the attached version.

You have to replace Http.php and Controller.php. After that it should be working again.

Second problem was caused by windows mobile sending no user-agent header for some requests.

I tested it with my test windows mobile phone and it is working now.

**Comment by Gloirin** [comment]
**Comment posted by schoebelh on 29 Nov 2014 13:46**

Solution proved! Thank you very much!
